server/vmap/trillian_map_server.go

```diff
@@ -74,33 +85,51 @@ func (t *TrillianMapServer) GetLeaves(ctx context.Context, req *trillian.GetMapL
 
 	smtReader := merkle.NewSparseMerkleTreeReader(req.Revision, hasher, tx)
 
-	leaves, err := tx.Get(ctx, req.Revision, req.Index)
-	if err != nil {
-		return nil, err
-	}
-	glog.Infof("%v: wanted %v leaves, found %v", mapID, len(req.Index), len(leaves))
+	inclusions := make([]*trillian.MapLeafInclusion, 0, len(req.Index))
+	found := 0
+	for _, index := range req.Index {
+		// TODO(gdbelvin): specify the index length in the tree specification.
+		if got, want := len(index), indexSize; got != want {
+			return nil, status.Errorf(codes.InvalidArgument,
+				"index len(%x): %v, want %v", index, got, want)
+		}
+		// Fetch the leaf if it exists.
+		leaves, err := tx.Get(ctx, req.Revision, [][]byte{index})
+		if err != nil {
+			return nil, err
+		}
+		var leaf *trillian.MapLeaf
+		if len(leaves) == 1 {
+			leaf = &leaves[0]
+			found++
+		} else {
+			// Empty leaf for proof of non-existence.
+			leaf = &trillian.MapLeaf{
+				Index:     index,
+				LeafValue: nil,
+			}
+		}
 
-	resp := &trillian.GetMapLeavesResponse{
-		MapLeafInclusion: make([]*trillian.MapLeafInclusion, len(leaves)),
-	}
-	for i, leaf := range leaves {
-		proof, err := smtReader.InclusionProof(ctx, req.Revision, leaf.Index)
+		// Fetch the proof regardless of whether the leaf exists.
```

Comment:
If the leaf does not exist, the proof will still be a a full merkle tree inclusion path. It's just that the leaf value will be nil. Hashing that with all the neighbor values up to the root will produce the correct root hash.